.../input/entityanalytics/provider/okta: Rate limiting fix, improvements

chrisberkhout opened the pull request:
## Proposed commit message

```
.../input/entityanalytics/provider/okta: Rate limiting fix, improvements

- Fix a bug in the stopping of requests when `x-rate-limit-remaining: 0`.
- Add a deadline so long waits return immediately as errors.
- Add an option to set a fixed request rate.
```

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

- Relates #40106

elasticmachine commented:
Pinging @elastic/security-service-integrations (Team:Security-Service Integrations)

efd6 commented on x-pack/filebeat/input/entityanalytics/provider/okta/conf.go:
```diff
@@ -27,6 +27,7 @@ func defaultConfig() conf {
 		SyncInterval:   24 * time.Hour,
 		UpdateInterval: 15 * time.Minute,
 		LimitWindow:    time.Minute,
+		LimitFixed:     nil,
```
This isn't needed. Is it here for explicitness?

chrisberkhout replied:
I guess I was thinking out loud. Doesn't need to be there. Removed.

efd6 commented on x-pack/filebeat/input/entityanalytics/provider/okta/internal/okta/ratelimiter.go:
```diff
@@ -16,33 +16,66 @@ import (
 )
 
-type RateLimiter map[string]*rate.Limiter
+type RateLimiter struct {
```
I think this needs godoc.

chrisberkhout replied:
Added.

efd6 commented on x-pack/filebeat/input/entityanalytics/provider/okta/internal/okta/ratelimiter.go:
```diff
@@ -16,33 +16,66 @@ import (
-	return r
+const waitDeadline = 30 * time.Minute
+
+func NewRateLimiter(window time.Duration, fixedLimit *int) *RateLimiter {
```
Godoc here, particularly for the parameter semantics.

chrisberkhout replied:
Added.

efd6 commented on x-pack/filebeat/input/entityanalytics/provider/okta/internal/okta/ratelimiter.go:
```diff
@@ -16,33 +16,66 @@ import (
+		limit = rate.Limit(float64(*r.fixedLimit) / r.window.Seconds())
+	}
+	limiter := rate.NewLimiter(limit, 1) // Allow a single fetch operation to obtain limits from the API
+	ready := make(chan struct{})
+	close(ready)
```
Suggest something like
```
// alwaysReady is an non-blocking chan.
var alwaysReady = make(chan struct{})

func init() { close(alwaysReady) }

func (r RateLimiter) endpoint(path string) endpointRateLimiter {
	if existing, ok := r.byEndpoint[path]; ok {
		return existing
	}
	limit := rate.Limit(1)
	if r.fixedLimit != nil {
		limit = rate.Limit(float64(*r.fixedLimit) / r.window.Seconds())
	}
	limiter := rate.NewLimiter(limit, 1) // Allow a single fetch operation to obtain limits from the API
	newEndpointRateLimiter := endpointRateLimiter{
		limiter: limiter,
		ready:   alwaysReady,
	}
	r.byEndpoint[path] = newEndpointRateLimiter
	return newEndpointRateLimiter
}
```
so that we have a single never-blocking chan.

chrisberkhout replied:
Done with the name `immediatelyReady`.
It's true that channel will always be ready, but the `endpointRateLimiter`'s `ready` field won't always have that value...
I don't know. Is this better, or the same? (or less good?)

efd6 commented on x-pack/filebeat/input/entityanalytics/provider/okta/internal/okta/ratelimiter.go:
```diff
@@ -16,33 +16,66 @@ import (
-	log.Debugw("rate limit", "limit", limiter.Limit(), "burst", limiter.Burst(), "url", url.String())
-	return limiter.Wait(ctx)
+	e := r.endpoint(endpoint)
+	<-e.ready
```
```suggestion
	<-e.ready
	select {
	case <-e.ready:
	case <-ctx.Done():
		return ctx.Err()
	}
```

chrisberkhout replied:
:+1:

efd6 commented on x-pack/filebeat/input/entityanalytics/provider/okta/internal/okta/ratelimiter.go:
```diff
@@ -16,33 +16,66 @@ import (
-func NewRateLimiter() RateLimiter {
-	r := make(RateLimiter)
-	return r
+const waitDeadline = 30 * time.Minute
```
Deadline conventionally is a `time.Time`. `waitTimeout`? I think this wants godoc too, to explain what its intention is.

efd6 commented on x-pack/filebeat/input/entityanalytics/provider/okta/internal/okta/ratelimiter.go:
```diff
@@ -16,33 +16,66 @@ import (
+	e := r.endpoint(endpoint)
+	<-e.ready
+	log.Debugw("rate limit", "limit", e.limiter.Limit(), "burst", e.limiter.Burst(), "url", url.String())
+	ctxWithDeadline, cancel := context.WithDeadline(ctx, time.Now().Add(waitDeadline))
```
Should the deadline be calculated before the `e.ready` recv?

chrisberkhout replied:
I intended to apply it only when using the rate limiter, not when requests are shut down, but I'll change it to apply to both.

chrisberkhout replied:
Done.

efd6 commented on x-pack/filebeat/input/entityanalytics/provider/okta/internal/okta/ratelimiter_test.go:
```diff
@@ -17,68 +19,145 @@ func TestRateLimiter(t *testing.T) {
+		target := 30.0
+		buffer := 0.01
+
+		if tokens < target-buffer || tokens > target+buffer {
```
```suggestion
		if tokens < target-buffer || target+buffer < tokens {
```

chrisberkhout replied:
Done.

efd6 commented on x-pack/filebeat/input/entityanalytics/provider/okta/internal/okta/ratelimiter_test.go:
```diff
@@ -17,68 +19,145 @@ func TestRateLimiter(t *testing.T) {
+		wait := time.Since(start)
+
+		if wait > 1010*time.Millisecond {
+			t.Errorf("doesn't allow requests to resume after reset. had to wait %d milliseconds", wait.Milliseconds())
```
```suggestion
			t.Errorf("doesn't allow requests to resume after reset. had to wait %s", wait)
```

chrisberkhout replied:
Done.

efd6 commented on x-pack/filebeat/input/entityanalytics/provider/okta/internal/okta/ratelimiter_test.go:
```diff
@@ -17,68 +19,145 @@ func TestRateLimiter(t *testing.T) {
+		r := NewRateLimiter(window, fixedLimit)
 		const endpoint = "/foo"
-		limiter := r.limiter(endpoint)
+		url, _ := url.Parse(endpoint)
```
```suggestion
		url, err := url.Parse(endpoint)
		if err != nil {
			t.Fatalf("failed to parse endpoint: %v", err)
		}
```

chrisberkhout replied:
Added.

efd6 commented on x-pack/filebeat/input/entityanalytics/provider/okta/internal/okta/ratelimiter_test.go:
```diff
@@ -17,68 +19,145 @@ func TestRateLimiter(t *testing.T) {
+
+		const endpoint = "/foo"
+
+		url, _ := url.Parse(endpoint)
```
Check error.

chrisberkhout replied:
Done.

chrisberkhout commented:
@efd6 Addressed comments. Also added documentation changes that I just forgot to commit before. Now also has `CHANGELOG.next.asciidoc` entries (1 bugfix, 1 addition/improvement).

efd6 reviewed (approved):
Thanks
